File: examples/AndroidMathUtility/Vector.cpp
#include <sstream>

#include "AndroidMathUtility.h"

namespace AndroidMathUtility {

Vector::Vector()
{
	this->size = 0;
	this->data = NULL;
}

Vector::Vector(int size)
{
	data = new float[size];
	for(int i = 0; i < size; i++) {
		data[i] = 0.0f;
	}
	this->size = size;
}

Vector::Vector(const Vector& vec)
{
	if(vec.GetSize() > 0) {
		this->size = vec.GetSize();
		this->data = new float[size];
		for(int i = 0; i < size; i++) {
			this->data[i] = vec.data[i];
		}
	}
	else {
		this->size = 0;
		this->data = NULL;
	}
}

Vector::~Vector()
{
	delete[] data;
}

float Vector::Get(int id) const
{
	return this->data[id];
}

int Vector::GetSize() const
{
	return this->size;
}

void Vector::Set(int id, float value)
{
	this->data[id] = value;
}

void Vector::MultipleByScalar(float scalar)
{
	for(int i = 0; i < size; i++) {
		this->data[i] *= scalar;
	}
}

std::string Vector::ToString() const
{
	std::stringstream ss;
	ss << "[";
	for(int i = 0; i < size; i++) {
		ss << data[i];
		if(i + 1 < size)
			ss << ", ";
	}
	ss << "]";
	return ss.str();
}

}

File: examples/AndroidMath/Vector.java
package org.upp.AndroidMath;

/**
 * Math vector class which whole implementation is native (c/c++).
 */
public class Vector
{
	/**
	 * Field use by MemoryManager to store pointer to c++ object.
	 * If you want to use MemoryManager with your class make sure you have this field
	 * in your class.
	 */
	private long nativeAdress = 0;
	
	/**
	 * In order to avoid memory leaks on the native side. The cleaning method needs to be
	 * called.
	 *
	 * In previous version of this example finalize method was used. However, in Java 9 the
	 * finalize method was deprecated and we can not longer relay on them.
	 */
	protected void dispose()
	{
		nativeDispose();
	}
	
	public Vector(int size)
	{
		construct(size);
	}
	
	public Vector(Vector vec)
	{
		copyConstruct(vec);
	}
	
	// Native stuff - C/C++
	public native int   getSize();
	public native float get(int id);
	
	public native void set(int id, float value);
	
	public native void multipleByScalar(float scalar);
	
	public native String toString();
	
	private native void construct(int size);
	private native void copyConstruct(Vector vec);
	private native void nativeDispose();
}
